test/cpp/api/inference_mode.cpp

```diff
@@ -0,0 +1,391 @@
+#include <torch/script.h>
+#include <gtest/gtest.h>
+#include <test/cpp/api/support.h>
+
+using namespace torch::autograd;
+using namespace torch::test;
+
+namespace {
+  torch::Tensor functional_op(torch::Tensor& x) {
+    return x + x;
+  }
+
+  void inplace_op(torch::Tensor& x) {
+    x.add_(1);
+  }
+
+  torch::Tensor view_op(torch::Tensor& x) {
+    return x.view({2, 3});
+  }
+}
+
+/*
+  Only the following combos of Autograd & InplaceOrView keys on tensors are valid:
+    - Autograd=true, InplaceOrView=true (normal tensor)
+    - Autograd=false, InplaceOrView=false (inference tensor)
+  Tensors created in InferenceMode are mostly inference tensors. There's an exception
+  which listed below, even though these tensors are created in InferenceMode, they're
+  normal tensors:
+    - tensors which are view of normal tensors.
+  Note output of inplace ops on normal tensors(which is the input itself).
```

Comment:
Not sure what this means. The last 3 lines here are a bit confusing (maybe remove the list if there is a single element in that list).